apollo-router/src/axum_factory/listeners.rs

```diff
@@ -496,6 +453,45 @@ pub(super) fn serve_router_on_listen_addr(
     (server, shutdown_sender)
 }
 
+/// Configures a connection, no matter whether it's TLS, TCP, or UDS (ie, unix sockets) with
+/// whatever parameters are configured by the end-user in their router config
+///
+/// NOTE: centralize all connection configuration changes here so that the behavior of the
+/// connections remains in-step
+fn configure_connection(
+    conn_builder: &mut Builder<TokioExecutor>,
+    header_read_timeout: Duration,
+    opt_max_http1_headers: Option<usize>,
+    opt_max_http1_buf_size: Option<ByteSize>,
+    opt_max_http2_headers_list_bytes: Option<ByteSize>,
+) -> Http1Builder<'_, TokioExecutor> {
+    // NOTE: this is a builder that auto-detects http1/http2, so we can configure both and rely on
+    // it to figure out whether we have an http1 or http2 connection
+    let mut builder = conn_builder.http2();
+    if let Some(max_header_list_size) = opt_max_http2_headers_list_bytes {
+        // ByteSize has a .as_u64(), but doesn't have a as_u32(), to explain
+        // this funky `as` conversion
+        builder.max_header_list_size(max_header_list_size.as_u64() as u32);
```

Comment:
Could it happen that the header size would be bigger than `u32::MAX` ? If so I would prefer using `u32::try_from()` instead of `as` cast. WDYT ?